trin-core/src/portalnet/find/iterators/findnodes.rs

```diff
@@ -0,0 +1,701 @@
+    pub num_results: usize,
+
+    /// The timeout for a single peer.
+    ///
+    /// If a successful result is not reported for a peer within this timeout
+    /// window, the iterator considers the peer unresponsive and will not wait for
+    /// the peer when evaluating the termination conditions, until and unless a
+    /// result is delivered. Defaults to `10` seconds.
+    pub peer_timeout: Duration,
+}
+
+impl<TNodeId> FindNodeQuery<TNodeId>
+where
+    TNodeId: Into<Key<TNodeId>> + Eq + Clone,
+{
+    /// Creates a new query with the given configuration.
+    pub fn with_config<I>(
+        config: FindNodeQueryConfig,
+        target_key: Key<TNodeId>,
+        known_closest_peers: I,
+    ) -> Self
+    where
+        I: IntoIterator<Item = Key<TNodeId>>,
+    {
+        // Initialise the closest peers to begin the query with.
+        let closest_peers = known_closest_peers
+            .into_iter()
+            .map(|key| {
+                let key: Key<TNodeId> = key;
+                let distance = key.distance(&target_key);
+                let state = QueryPeerState::NotContacted;
+                (distance, QueryPeer::new(key, state))
+            })
+            .take(config.num_results)
+            .collect();
+
+        // The query initially makes progress by iterating towards the target.
+        let progress = QueryProgress::Iterating { no_progress: 0 };
+
+        FindNodeQuery {
+            config,
+            target_key,
+            progress,
+            closest_peers,
+            num_waiting: 0,
+        }
+    }
+
+    /// Callback for delivering the result of a successful request to a peer
+    /// that the query is waiting on.
+    ///
+    /// Delivering results of requests back to the query allows the query to make
+    /// progress. The query is said to make progress either when the given
+    /// `closer_peers` contain a peer closer to the target than any peer seen so far,
+    /// or when the query did not yet accumulate `num_results` closest peers and
+    /// `closer_peers` contains a new peer, regardless of its distance to the target.
+    ///
+    /// After calling this function, `next` should eventually be called again
+    /// to advance the state of the query.
+    ///
+    /// If the query is finished, the query is not currently waiting for a
+    /// result from `peer`, or a result for `peer` has already been reported,
+    /// calling this function has no effect.
+    pub fn on_success(&mut self, node_id: &TNodeId, closer_peers: Vec<TNodeId>) {
+        if let QueryProgress::Finished = self.progress {
+            return;
+        }
+
+        let key: Key<TNodeId> = node_id.clone().into();
+        let distance = key.distance(&self.target_key);
+
+        // Mark the peer's progress, the total nodes it has returned and it's current iteration.
+        // If the node returned peers, mark it as succeeded.
+        match self.closest_peers.entry(distance) {
```

Comment:
IIUC, this is true for the XOR distance metric, but not for the state distance function.

Reply:
Good point @njgheorghita. Added a note. After thinking about it, I think it's extremely unlikely that it does happen (with XOR) but even if it does, it doesn't look like the results would be very severe if presented with results from a node that it thinks it already got results from. It would just return, so basically one of the nodes would be chosen and the other ignored.

Good point @jacobkaufmann. I'll add that to the issue I file about making queries distance function configurable.